### PR TITLE
feat: [PIE-6201]: NG - Also save webhook secret when registering webhook 

### PR DESCRIPTION
**Pull request by vcalasansh**
## Describe your changes
This PR adds webhook secret to webhook creation request payload when registering webhooks in git repo providers. This is currently only supported for GitHub webhooks.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41537)

### CODE REVIEW

**Comment by vcalasansh** [comment]
trigger smartchecks

**Comment by vcalasansh** [comment]
trigger smartchecks

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 934 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNHyyM6IxzAnQL4dqQ:-NLNHyyM6IxzAnQL4dqR:b-tz259h) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L934)):*
> ```Java
>     WebhookSecretData webhookSecretData = null;
>     String secretIdentifierRef = upsertWebhookRequestDTO.getWebhookSecretIdentifierRef();
>     if (isNotEmpty(secretIdentifierRef)) {
> ```

is it expected in any case that secretIdentifierRef = null? If not, then we should throw exception right away in such a case instead of returning null

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 598 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIF6k034txqHARskc:-NLNIF6k034txqHARskd:bsmn9z1) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L598)):*
> ```Java
>         getEncryptedDataDetails(upsertWebhookRequestDTO.getAccountIdentifier(),
>             upsertWebhookRequestDTO.getOrgIdentifier(), upsertWebhookRequestDTO.getProjectIdentifier(), scmConnector);
>     final WebhookSecretData webhookSecretData = getWebhookSecretData(upsertWebhookRequestDTO);
> ```

This will impact all flows that register webhooks, including old git-sync flows. So, be careful about the impact.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 401 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIaF3B3ttucBkhZDO:-NLNIaF3B3ttucBkhZDP:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L401)):*
> ```Java
>     String webhookSecret = null;
>     String secretIdentifierRef = upsertWebhookRequestDTO.getWebhookSecretIdentifierRef();
>     if (isNotEmpty(secretIdentifierRef)) {
> ```

same as above

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 422 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIjuQ6gaL6Ak8Nd69:-NLNIjuQ6gaL6Ak8Nd6A:b-gyhufv) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L422)):*
> ```Java
>             String.valueOf(webhookSecretData.getWebhookEncryptedSecretDTO().getSecretRef().getDecryptedValue());
>       } catch (RuntimeException ex) {
>         log.error("Error finding secret key in secret manager", ex);
> ```

why are we consuming the exception here itself instead of throwing it back?

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmGitWebhookTask.java` line 67 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIt_T3y26r-5mAJID:-NLNIt_T3y26r-5mAJIE:b-pv1hmw) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmGitWebhookTask.java#L67)):*
> ```Java
>         scmGitWebhookTaskParams.gitWebhookDetails.setSecret(webhookSecret);
>       } catch (RuntimeException ex) {
>         log.error("Error finding secret key in secret manager", ex);
> ```

is it fine to consume the exception here without any further action?

**Review by vcalasansh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41537)* status: 11 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 598 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIF6k034txqHARskc:-NLTxbIQ2PkiYDZ2pkdU:b-90j5el) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L598)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This will impact all flows that register webhooks, including old git-sync flows. So, be careful about the impact.
</blockquote></details>

Thanks for reviewing! Since the `webhookSecretIdentifierRef` (which is consumed, if present, by the `getWebhookSecret` method) is being added as a new field to the `UpsertWebhookRequestDTO` class, I believe there would be no danger or impact to anything that was previously working, right?
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 934 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNHyyM6IxzAnQL4dqQ:-NLTyAT-2qPEhi_AgDKc:b-dqwwno) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L934)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

is it expected in any case that secretIdentifierRef = null? If not, then we should throw exception right away in such a case instead of returning null
</blockquote></details>

Yes, almost all cases will have `secretIdentifierRef` as null. It will only be not-null for Github webhooks that user wants to use authentication with. Then they will provide a secret upon webhook creation, and (after this PR is merged) this secret will be registered in the webhook's config at Github's side. Thank you for the comment!
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 401 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIaF3B3ttucBkhZDO:-NLTyt_rA8j2pOURiZ5i:bc0wc0g) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L401)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Please consider my comment above.
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 422 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIjuQ6gaL6Ak8Nd69:-NLTzZAJ8ncBqK6NcNWV:b-aga6vk) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L422)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

why are we consuming the exception here itself instead of throwing it back?
</blockquote></details>

I thought it would be better to not make the entire webhook registration flow fail in case there is some issue with communicating with secret manager or decrypting the secret. The user can always go and add the secret manually in GitHub's side later. But this is arguable, please let me know if you think it would be best to throw an error here instead. Thanks for highlighting this!
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmGitWebhookTask.java` line 67 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIt_T3y26r-5mAJID:-NLU-Rji3LiZN7ehgI8N:bhpnnz1) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmGitWebhookTask.java#L67)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

is it fine to consume the exception here without any further action?
</blockquote></details>

Arguably yes, same as above. The user can always go and add the secret manually to the webhook config in Github's side.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 598 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIF6k034txqHARskc:-NLUO_csDAmzgph6Q0vv:b-xboybn) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L598)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Thanks for reviewing! Since the `webhookSecretIdentifierRef` (which is consumed, if present, by the `getWebhookSecret` method) is being added as a new field to the `UpsertWebhookRequestDTO` class, I believe there would be no danger or impact to anything that was previously working, right?
</blockquote></details>

Agreed. It depends on how we are using "webhookSecretData" further in our flows. Approving.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 422 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIjuQ6gaL6Ak8Nd69:-NLUPG3s7UH9Q8UkeLhy:b-rgdarp) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L422)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

I thought it would be better to not make the entire webhook registration flow fail in case there is some issue with communicating with secret manager or decrypting the secret. The user can always go and add the secret manually in GitHub's side later. But this is arguable, please let me know if you think it would be best to throw an error here instead. Thanks for highlighting this!
</blockquote></details>

So, the thing is as per what you explained if user wants authentication then this flow kicks in. And if we register it without authentication without even informing the user, that will be wrong. This should be thought upon properly on whats the expectation here.

**Review by vcalasansh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41537)* status: 11 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 422 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIjuQ6gaL6Ak8Nd69:-NLW7B9O5AaDaTqT4yaO:brjpjrx) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L422)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

So, the thing is as per what you explained if user wants authentication then this flow kicks in. And if we register it without authentication without even informing the user, that will be wrong. This should be thought upon properly on whats the expectation here.
</blockquote></details>

Thanks for the discussion, man! Upon a second thought, I agree with your concern. I will change this behavior: if secret decryption fails, this will throw an informative error message. I'll let you know once the PR is updated.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java` line 422 at r1](https://reviewable.io/reviews/harness/harness-core/41537#-NLNIjuQ6gaL6Ak8Nd69:-NLbXTON8mTLq0UCcxf7:b-kbuicr) ([raw file](https://github.com/harness/harness-core/blob/ced4dfd1ab6e886aeb4eba60e853d57c1b8cf023/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmManagerFacilitatorServiceImpl.java#L422)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Thanks for the discussion, man! Upon a second thought, I agree with your concern. I will change this behavior: if secret decryption fails, this will throw an informative error message. I'll let you know once the PR is updated.
</blockquote></details>

Approving it. Another suggestion would be to add another explanation/hint exception where this method is used so that it captures a more related message based on use-case using this method. eg: Lets say if connector creation called this method and got any exception, then it can add an explanationException on top of it saying "Connection creation failed due to failure in processing webhook secret" or similar. This gives more details to the customer.

We need to think if what we are sending back right now would be of any use to the user and if thats good enough for the user to understand the next steps.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 8 of 11 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41537)* status: 12 file(s) to review. 1 of 2 approvals obtained.

**Review by sribalij — APPROVED** [review]
> Reviewed 8 of 11 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41537)* status: :shipit: complete! ✓

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/scm/ScmGitWebhookTaskParams.java [hunk 1]
```diff
@@ -18,20 +18,22 @@
...
+import io.harness.ngtriggers.WebhookSecretData;
 import io.harness.security.encryption.EncryptedDataDetail;
 
 import java.util.List;
 import lombok.Builder;
-import lombok.Value;
+import lombok.Data;
 
-@Value
+@Data
```
Comment (vcalasansh) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/scm/ScmGitWebhookTaskParams.java:28]: This

**Pull request merged**